forest/benchmarking/operator_tools/random_operators.py

```diff
@@ -0,0 +1,209 @@
+    fiducial_vec[0] = 1
+    return np.matmul(unitary, fiducial_vec)
+
+
+def ginibre_state_matrix(dim: int, rank: int) -> np.ndarray:
+    """
+    Given a Hilbert space dimension dim and a desired rank K, returns a dim by dim positive
+    semidefinite matrix of rank K drawn from the Ginibre ensemble. For dim = K these are states
+    drawn from the Hilbert-Schmidt measure.
+
+    See reference [IM] for more details.
+
+    :param dim: Hilbert space dimension.
+    :param rank: The rank of a state.
+    :return: Returns a dim by rank matrix, drawn from the Ginibre ensemble.
+    """
+    if rank > dim:
+        raise ValueError("The rank of the state matrix cannot exceed the dimension.")
+    A = ginibre_matrix_complex(dim, rank)
+    M = A.dot(np.transpose(np.conjugate(A)))
+    return M / np.trace(M)
+
+
+def bures_measure_state_matrix(dim: int) -> np.ndarray:
+    """
+    Given a Hilbert space dimension dim, returns a dim by dim positive semidefinite matrix drawn
+    from the Bures measure.
+
+    [OSZ] Random Bures mixed states and the distribution of their purity
+          Osipov et al.,
+          J. Phys. A: Math. Theor. 43, 055302 (2010).
+          https://doi.org/10.1088/1751-8113/43/5/055302
+          https://arxiv.org/abs/0909.5094
+
+    :param dim: Hilbert space dimension.
+    :return: Returns a dim by dim matrix, drawn from the Bures measure.
+    """
+    A = ginibre_matrix_complex(dim, dim)
+    U = haar_rand_unitary(dim)
+    Udag = np.transpose(np.conjugate(U))
+    Id = np.eye(dim)
+    M = A.dot(np.transpose(np.conjugate(A)))
+    P = (Id + U).dot(M).dot(Id + Udag)
+    return P / np.trace(P)
+
+
+def rand_map_with_BCSZ_dist(dim: int, kraus_rank: int) -> np.ndarray:
+    """
+    Given a Hilbert space dimension dim and a Kraus rank K, returns a $dim^2 by dim^2$ Choi
+    matrix $J(Λ)$ of a channel drawn from the BCSZ distribution with Kraus rank $K$.
+
+    [RQO] Random quantum operations,
+          Bruzda et al.,
+          Physics Letters A 373, 320 (2009).
+          https://doi.org/10.1016/j.physleta.2008.11.043
+          https://arxiv.org/abs/0804.2361
+    
+    :param dim: Hilbert space dimension.
+    :param kraus_rank: The number of Kraus operators in the operator sum description of the channel.
+    :return: dim^2 by dim^2 Choi matrix, drawn from the BCSZ distribution with Kraus rank K.
+    """
+    # TODO: this ^^ is CPTP, might want a flag that allows for just CP quantum operations.
+    X = ginibre_matrix_complex(dim ** 2, kraus_rank)
+    rho = X @ X.conj().T
+    rho_red = partial_trace(rho, [0], [dim, dim])
+    # Note that Eqn. 8 of [RQO] uses a *row* stacking convention so in that case we would write
+    # Q = np.kron(np.eye(D), sqrtm(la.inv(rho_red)))
+    # But as we use column stacking we need:
+    Q = np.kron(sqrtm(la.inv(rho_red)), np.eye(dim))
+    Z = Q @ rho @ Q
+    return Z
+
+
+def permute_tensor_factors(dims: Union[int, List[int]], perm: List[int]) -> np.ndarray:
```

Comment:
Should we move this to somewhere else? It doesn't really fit here.

Reply:
Ok lets make a PR to put this somewhere else.